backend/src/schema/resolvers/searches.js

```diff
@@ -0,0 +1,74 @@
+import log from './helpers/databaseLogger'
+
+export default {
+  Query: {
+    findResources: async (_parent, args, context, _resolveInfo) => {
+      const { query, limit } = args
+      const { id: thisUserId } = context.user
+      // see http://lucene.apache.org/core/8_3_1/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#package.description
+      const myQuery = query
+        .replace(/\s+/g, ' ')
+        .replace(/[[@#:*~\\$|^\]?/"'(){}+?!,.-;]/g, '')
+        .split(' ')
+        .map(s => (s.toLowerCase().match(/^(not|and|or)$/) ? '"' + s + '"' : s + '*'))
+        .join(' ')
+      const postCypher = `
+      CALL db.index.fulltext.queryNodes('post_fulltext_search', $query)
+      YIELD node as resource, score
+      MATCH (resource)<-[:WROTE]-(author:User)
+      WHERE score >= 0.5
+      AND NOT (
+        author.deleted = true OR author.disabled = true
+        OR resource.deleted = true OR resource.disabled = true
+        OR (:User { id: $thisUserId })-[:BLOCKED]-(author)
+      )
+      WITH resource, author,
+      [(resource)<-[:COMMENTS]-(comment:Comment) | comment] as comments,
+      [(resource)<-[:SHOUTED]-(user:User) | user] as shouter
+      RETURN resource {
+        .*,
+        __typename: labels(resource)[0],
+        author: properties(author),
+        commentsCount: toString(size(comments)),
+        shoutedCount: toString(size(shouter))
+      }
+      LIMIT $limit
+      `
+
+      const userCypher = `
+      CALL db.index.fulltext.queryNodes('user_fulltext_search', $query)
+      YIELD node as resource, score
+      MATCH (resource)
+      WHERE score >= 0.5
+      AND NOT (resource.deleted = true OR resource.disabled = true
+      OR (:User { id: $thisUserId })-[:BLOCKED]-(resource))
+      RETURN resource {.*, __typename: labels(resource)[0]}
+      LIMIT $limit
+      `
+
+      const session = context.driver.session()
+      const searchResultPromise = session.readTransaction(async transaction => {
+        const postTransactionResponse = transaction.run(postCypher, {
+          query: myQuery,
+          limit,
+          thisUserId,
+        })
+        const userTransactionResponse = transaction.run(userCypher, {
+          query: myQuery,
+          limit,
+          thisUserId,
+        })
+        return Promise.all([postTransactionResponse, userTransactionResponse])
+      })
+
+      try {
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-20T16:01:15Z" title="Friday, December 20th 2019, 5:01:15 pm +01:00">Dec 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,69 @@
+const transformReturnType = record => {
+  return {
+    __typename: record.get('type'),
+    ...record.get('resource').properties,
+  }
+}
+export default {
+  Query: {
+    findResources: async (_parent, args, context, _resolveInfo) => {
+      const { query, limit } = args
+      const { id: thisUserId } = context.user
+      // see http://lucene.apache.org/core/8_3_1/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#package.description
+      const myQuery = query + '*'
+      const postCypher = `
+      CALL db.index.fulltext.queryNodes('post_fulltext_search', $query)
+      YIELD node as resource, score
+      MATCH (resource)<-[:WROTE]-(user:User)
+      WHERE score >= 0.5
+      AND NOT (user.deleted = true OR user.disabled = true
+      OR resource.deleted = true OR resource.disabled = true
+      OR (:User { id: $thisUserId })-[:BLOCKED]-(user))
+      RETURN resource, labels(resource)[0] AS type
+      LIMIT $limit
+      `
+      const session = context.driver.session()
+      let postResults, userResults
+      const readPostTxResultPromise = session.readTransaction(async transaction => {
+        const postTransactionResponse = transaction.run(postCypher, {
+          query: myQuery,
+          limit,
+          thisUserId,
+        })
+        return postTransactionResponse
+      })
+      try {
+        postResults = await readPostTxResultPromise
+      } finally {
+        session.close()
+      }
+
+      const userCypher = `
+      CALL db.index.fulltext.queryNodes('user_fulltext_search', $query)
+      YIELD node as resource, score
+      MATCH (resource)
+      WHERE score >= 0.5
+      AND NOT (resource.deleted = true OR resource.disabled = true
+      OR (:User { id: $thisUserId })-[:BLOCKED]-(resource))
+      RETURN resource, labels(resource)[0] AS type
+      LIMIT $limit
+      `
+      const readUserTxResultPromise = session.readTransaction(async transaction => {
+        const userTransactionResponse = transaction.run(userCypher, {
+          query: myQuery,
+          limit,
+          thisUserId,
+        })
+        return userTransactionResponse
+      })
+      try {
+        userResults = await readUserTxResultPromise
+      } finally {
+        session.close()
+      }
+      let searchResults = [...postResults.records, ...userResults.records]
```


~If we're appending users to posts and not sorting them in one array by relevance, I would prefer a graphql query like this:~
```graphql
{
  search(query: "Foobar") {
    posts {
      title
    }
    users {
      name
    }
}
```
The search results are sorted by relevance by default (highest score first).

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-20T16:42:34Z" title="Friday, December 20th 2019, 5:42:34 pm +01:00">Dec 20, 2019</time>_

---


Actually, I changed my mind. After looking through the code it seems i would break a lot with an API like this :point_up: 

@Mogge would it be ok if we remove `findPosts` and `findUsers` instead? We're not using anywhere, or do we?